src/sentry/issues/escalating.py

```diff
@@ -0,0 +1,68 @@
+from datetime import datetime, timedelta
+from typing import List, Tuple
+
+from snuba_sdk import (
+    Column,
+    Condition,
+    Direction,
+    Entity,
+    Function,
+    Limit,
+    Offset,
+    Op,
+    OrderBy,
+    Query,
+    Request,
+)
+
+from sentry.models import Group
+from sentry.utils.json import JSONData
+from sentry.utils.snuba import raw_snql_query
+
+
+def extract_project_and_group_ids(groups: List[Group]) -> Tuple[List[int], List[int]]:
+    """Return all project and group IDs from a list of Group"""
+    group_ids = []
+    project_ids = []
+    for group in groups:
+        project_ids.append(group.project_id)
+        group_ids.append(group.id)
+
+    # This also removes duplicated project ids
+    return list(set(project_ids)), group_ids
+
+
+def query_groups_past_counts(groups: List[Group]) -> JSONData:
+    project_ids, group_ids = extract_project_and_group_ids(groups)
+    group_id_col = Column("group_id")
+    proj_id_col = Column("project_id")
+
+    query = Query(
+        match=Entity("events"),
+        select=[
+            proj_id_col,
+            group_id_col,
+            Function("toStartOfHour", [Column("timestamp")], "hourBucket"),
+            Function("count", []),
+        ],
+        groupby=[proj_id_col, group_id_col, Column("hourBucket")],
+        where=[
+            Condition(proj_id_col, Op.IN, Function("tuple", project_ids)),
+            Condition(Column("group_id"), Op.IN, Function("tuple", group_ids)),
+            Condition(Column("timestamp"), Op.GTE, datetime.now() - timedelta(days=14)),
+            Condition(Column("timestamp"), Op.LT, datetime.today()),
+        ],
+        limit=Limit(10),  # TODO: Limit(10000)
+        offset=Offset(0),
+        orderby=[
+            # Make sure all the groups related to same project are returned first
+            # IIUC this will help Snuba with performance
+            OrderBy(proj_id_col, Direction.ASC),
+            OrderBy(group_id_col, Direction.ASC),
+            OrderBy(Column("hourBucket"), Direction.ASC),
```

Comment:
@onewland Is it more optimal to use `DESC` or `ASC`?

Do the orderbys make sense to you?

Reply:
I think what would help Snuba with performance would be to batch the input group_ids that you pass to lines 50 and 51. You have to handle the case where Snuba cannot take every archived group ID in a single query which means that `query_groups_past_counts` needs to either slice up its input or whatever calls `query_groups_past_counts` needs to slice up its input. The reason we recommend requesting group IDs from the same projects as much as possible is that ClickHouse can more quickly serve you rows or aggregates of rows that come from the block or blocks located nearby. The project_id is part of the ordering key that determines which rows get assigned to which blocks. If you send us input with 500 different `project_id`s it's much more likely we'll have to load 500 non-adjacent blocks then if you can provide us an input with 2 `project_id`s. You obviously don't have total control over this but as you're determining how to batch your queries this is something to keep in mind.

The ORDER BY in the query is mostly going to tell Snuba how to order its results. It's useful for iterating over resultsets that are >10k rows (because it makes passing the offset to LIMIT possible) but ORDER BY won't have much of an impact in your query performance because your filter already *should* be filtering down to a very small number of rows.

Reply:
I will handle the optimizations in a follow up task (filed). I want to merge this tomorrow.